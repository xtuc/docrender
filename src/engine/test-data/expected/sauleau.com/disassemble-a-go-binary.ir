(
    content: [
        Clear,
        Title((
            level: 2,
            text: Text("Simple program"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("The example project uses the following files:"),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Title((
            level: 3,
            text: Text("secret/secret.go:"),
        ), []),
        Clear,
        Paragraph([]),
        Code("go", "package secret\n\nimport \"log\"\n\nfunc ILoveTrains() {\n    log.Println(\"secret\")\n}\n"),
        Clear,
        Paragraph([]),
        Title((
            level: 3,
            text: Text("cmd/test/main.go:"),
        ), []),
        Clear,
        Paragraph([]),
        Code("go", "package main\n\nimport \"github.com/xtuc/example/secret\"\n\nfunc main() {\n    secret.ILoveTrains()\n}\n"),
        Clear,
        Title((
            level: 2,
            text: Text("Disassembly"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("You actually don\'t need to disassemble the binary. For the sake of simplicity, I\'ll only use strings against the binary."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("bash", "$ cd cmd/test\n$ go build\n$ strings ./test | grep secret\n\ngithub.com/xtuc/example/secret.ILoveTrains\n[…]\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("Note that Go won\'t include can that you don\'t use. All the imports are used."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("We can clearly see that github.com/xtuc is loving trains and even see the file structure. It makes no sense to include such information in the final binary, it\'s what we call a disclosure."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("You might have noticed that I didn\'t stripped the binary. Let\'s try again with a stripped binary."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("bash", "$ cd cmd/test\n$ go build -ldflags=\"-s -w\"\n$ file ./test\ntest: ELF 64-bit LSB executable […] stripped\n$ strings ./test | grep secret\n\ngithub.com/xtuc/example/secret.ILoveTrains\n[…]\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("Even with a stripped (it\'s meant for distribution) we can still see them."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Since the binary is stripped we don\'t have debugging symbols anymore:"),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("bash", "$ nm ./test\nnm: ./test: no symbols\n"),
        Clear,
        Title((
            level: 2,
            text: Text("Stack Overflow, what\'s the solution?"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("Stack Overflow can be useful in many cases, but not this time. This is the answer I read:"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Why do you need even that?"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Don\'t distribute Golang binaries."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Definitely not helpful."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Title((
            level: 2,
            text: Text("Golang, what\'s the solution?"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("I didn\'t ask, so this is not their answer: as many things in Golang, it\'s highly opinionated and specific to Google\'s needs."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Since they don\'t distribute any Go binary, why would they invest time in an obfuscation process."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Title((
            level: 2,
            text: Text("What\'s the real solution?"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("Since the Golang compiler won\'t mangle function names, I decided to mangle them before passing it to the compiler."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("It\'s a source mangler: "),
            ], false),
            InlineText([
                TextWithOptions("https://github.com/xtuc/manglo", (
                    link: Some("https://github.com/xtuc/manglo"),
                    styles: [
                        ("text-decoration", "underline"),
                        ("color", "rgb(6.666667%, 33.333336%, 80%)"),
                    ],
                )),
            ], false),
            InlineText([
                Text("."),
            ], false),
        ]),
        Paragraph([]),
        Paragraph([]),
        Title((
            level: 2,
            text: Text("Alternative compiler backends"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("My example uses the default Go compiler. I tried the gccgo compiler (GCC with a Go fronted) but it refused to compile because the imports weren\'t quite the same (might have changed somewhere around Go1.9)."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("GCC mangles all names."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Title((
            level: 2,
            text: Text("Packing?"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("Packing the binary will indeed removing the clear strings."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Anyone having disassemble skills will notice it immediately and revert the process (unpack)."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("I\'m not sure if it works cross-platform."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Paragraph([]),
    ],
    styles: {
        "pbNORMAL_TEXT": [
            ("line-height", "1.7249999"),
        ],
    },
)