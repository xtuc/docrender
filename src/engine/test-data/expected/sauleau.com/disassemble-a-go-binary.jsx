import { createElement, Fragment } from "https://unpkg.com/es-react";

function Heading({ style, id, level, children }) {
  const tag = "h" + level;
  return createElement(tag, { style, id }, children);
}

function Paragraph({ children }) {
  return createElement("p", {}, children);
}

function Code({ children }) {
  return createElement("code", {}, children);
}

function Image({ src }) {
  return createElement("img", { src });
}

function Link({ href }) {
  return createElement("a", { href });
}

function List({ children }) {
  return createElement("ul", {}, children);
}

function ListItem({ children }) {
  return createElement("li", {}, children);
}

export default function Doc({
  heading = Heading,
  paragraph = Paragraph,
  code = Code,
  image = Image,
  link = Link,
  list = List,
  listItem = ListItem,
} = {}) {
  return createElement(
    Fragment,
    {},
    createElement(heading, { level: 2, id: "pb-todo" }, `Simple program`),
    createElement(
      paragraph,
      {},
      `The example project uses the following files:`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(heading, { level: 3, id: "pb-todo" }, `secret/secret.go:`),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-go" },
      `<codeblock-js>package secret

import "log"

func ILoveTrains() {
    log.Println("secret")
}
</codeblock-js>`,
    ),
    createElement(paragraph, {}),
    createElement(heading, { level: 3, id: "pb-todo" }, `cmd/test/main.go:`),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-go" },
      `<codeblock-js>package main

import "github.com/xtuc/example/secret"

func main() {
    secret.ILoveTrains()
}
</codeblock-js>`,
    ),
    createElement(heading, { level: 2, id: "pb-todo" }, `Disassembly`),
    createElement(
      paragraph,
      {},
      `You actually don't need to disassemble the binary. For the sake of simplicity, I'll only use strings against the binary.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-bash" },
      `<codeblock-js>$ cd cmd/test
$ go build
$ strings ./test | grep secret

github.com/xtuc/example/secret.ILoveTrains
[…]
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `Note that Go won't include can that you don't use. All the imports are used.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `We can clearly see that github.com/xtuc is loving trains and even see the file structure. It makes no sense to include such information in the final binary, it's what we call a disclosure.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `You might have noticed that I didn't stripped the binary. Let's try again with a stripped binary.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-bash" },
      `<codeblock-js>$ cd cmd/test
$ go build -ldflags="-s -w"
$ file ./test
test: ELF 64-bit LSB executable […] stripped
$ strings ./test | grep secret

github.com/xtuc/example/secret.ILoveTrains
[…]
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `Even with a stripped (it's meant for distribution) we can still see them.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Since the binary is stripped we don't have debugging symbols anymore:`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-bash" },
      `<codeblock-js>$ nm ./test
nm: ./test: no symbols
</codeblock-js>`,
    ),
    createElement(
      heading,
      { level: 2, id: "pb-todo" },
      `Stack Overflow, what's the solution?`,
    ),
    createElement(
      paragraph,
      {},
      `Stack Overflow can be useful in many cases, but not this time. This is the answer I read:`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Why do you need even that?`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Don't distribute Golang binaries.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Definitely not helpful.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      heading,
      { level: 2, id: "pb-todo" },
      `Golang, what's the solution?`,
    ),
    createElement(
      paragraph,
      {},
      `I didn't ask, so this is not their answer: as many things in Golang, it's highly opinionated and specific to Google's needs.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Since they don't distribute any Go binary, why would they invest time in an obfuscation process.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      heading,
      { level: 2, id: "pb-todo" },
      `What's the real solution?`,
    ),
    createElement(
      paragraph,
      {},
      `Since the Golang compiler won't mangle function names, I decided to mangle them before passing it to the compiler.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `It's a source mangler: `,
      createElement(
        link,
        {
          style: {
            textDecoration: "underline",
            color: "rgb(6.666667%, 33.333336%, 80%)",
          },
          href: "https://github.com/xtuc/manglo",
        },
        `https://github.com/xtuc/manglo`,
      ),
      `.`,
    ),
    createElement(paragraph, {}),
    createElement(paragraph, {}),
    createElement(
      heading,
      { level: 2, id: "pb-todo" },
      `Alternative compiler backends`,
    ),
    createElement(
      paragraph,
      {},
      `My example uses the default Go compiler. I tried the gccgo compiler (GCC with a Go fronted) but it refused to compile because the imports weren't quite the same (might have changed somewhere around Go1.9).`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `GCC mangles all names.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(heading, { level: 2, id: "pb-todo" }, `Packing?`),
    createElement(
      paragraph,
      {},
      `Packing the binary will indeed removing the clear strings.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Anyone having disassemble skills will notice it immediately and revert the process (unpack).`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `I'm not sure if it works cross-platform.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(paragraph, {}),
  );
}
