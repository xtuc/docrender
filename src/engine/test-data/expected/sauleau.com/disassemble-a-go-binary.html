<style>
  .pbNORMAL_TEXT {
    line-height: 1.7249999;
  }
</style>
<h2 id="pb-Simple%20program">Simple program</h2>
<p>The example project uses the following files:</p>
<h3 id="pb-secret%2Fsecret.go%3A">secret/secret.go:</h3>
<pre><code class="language-go"><codeblock-js>package secret

import "log"

func ILoveTrains() {
    log.Println("secret")
}
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h3 id="pb-cmd%2Ftest%2Fmain.go%3A">cmd/test/main.go:</h3>
<pre><code class="language-go"><codeblock-js>package main

import "github.com/xtuc/example/secret"

func main() {
    secret.ILoveTrains()
}
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h2 id="pb-Disassembly">Disassembly</h2>
<p>
  You actually don't need to disassemble the binary. For the sake of simplicity,
  I'll only use strings against the binary.
</p>
<pre><code class="language-bash"><codeblock-js>$ cd cmd/test
$ go build
$ strings ./test | grep secret

github.com/xtuc/example/secret.ILoveTrains
[…]
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>
  Note that Go won't include can that you don't use. All the imports are used.
</p>
<p>
  We can clearly see that github.com/xtuc is loving trains and even see the file
  structure. It makes no sense to include such information in the final binary,
  it's what we call a disclosure.
</p>
<p>
  You might have noticed that I didn't stripped the binary. Let's try again with
  a stripped binary.
</p>
<pre><code class="language-bash"><codeblock-js>$ cd cmd/test
$ go build -ldflags="-s -w"
$ file ./test
test: ELF 64-bit LSB executable […] stripped
$ strings ./test | grep secret

github.com/xtuc/example/secret.ILoveTrains
[…]
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>Even with a stripped (it's meant for distribution) we can still see them.</p>
<p>Since the binary is stripped we don't have debugging symbols anymore:</p>
<pre><code class="language-bash"><codeblock-js>$ nm ./test
nm: ./test: no symbols
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h2 id="pb-Stack%20Overflow%2C%20what%27s%20the%20solution%3F">
  Stack Overflow, what's the solution?
</h2>
<p>
  Stack Overflow can be useful in many cases, but not this time. This is the
  answer I read:
</p>
<p>Why do you need even that?</p>
<p>Don't distribute Golang binaries.</p>
<p>Definitely not helpful.</p>
<h2 id="pb-Golang%2C%20what%27s%20the%20solution%3F">
  Golang, what's the solution?
</h2>
<p>
  I didn't ask, so this is not their answer: as many things in Golang, it's
  highly opinionated and specific to Google's needs.
</p>
<p>
  Since they don't distribute any Go binary, why would they invest time in an
  obfuscation process.
</p>
<h2 id="pb-What%27s%20the%20real%20solution%3F">What's the real solution?</h2>
<p>
  Since the Golang compiler won't mangle function names, I decided to mangle
  them before passing it to the compiler.
</p>
<p>
  It's a source mangler:
  <a
    style="text-decoration: underline; color: rgb(6.666667%, 33.333336%, 80%)"
    href="https://github.com/xtuc/manglo"
    >https://github.com/xtuc/manglo</a
  >.
</p>
<h2 id="pb-Alternative%20compiler%20backends">Alternative compiler backends</h2>
<p>
  My example uses the default Go compiler. I tried the gccgo compiler (GCC with
  a Go fronted) but it refused to compile because the imports weren't quite the
  same (might have changed somewhere around Go1.9).
</p>
<p>GCC mangles all names.</p>
<h2 id="pb-Packing%3F">Packing?</h2>
<p>Packing the binary will indeed removing the clear strings.</p>
<p>
  Anyone having disassemble skills will notice it immediately and revert the
  process (unpack).
</p>
<p>I'm not sure if it works cross-platform.</p>
